staging/src/k8s.io/externaljwt/pkg/plugin/keycache.go

```diff
@@ -0,0 +1,215 @@
+	client externaljwtv1alpha1.ExternalJWTSignerClient
+
+	syncGroup     singleflight.Group
+	listenersLock sync.Mutex
+	listeners     []externaljwt.Listener
+
+	// last successful refresh hint expires at this time.
+	nextRefreshHint atomic.Pointer[time.Time]
+	publicKeys      atomic.Pointer[VerificationKeys]
+}
+
+// newKeyCache constructs an implementation of KeyCache.
+func newKeyCache(client externaljwtv1alpha1.ExternalJWTSignerClient) *keyCache {
+	cache := &keyCache{
+		client: client,
+	}
+	return cache
+}
+
+// InitialFill can be used to performs an initial fetch for keys get the
+// refresh interval as recommended by external signer.
+func (p *keyCache) initialFill(ctx context.Context) error {
+	if err := p.syncKeys(ctx); err != nil {
+		return fmt.Errorf("while performing initial cache fill: %w", err)
+	}
+	return nil
+}
+
+func (p *keyCache) scheduleSync(ctx context.Context, keySyncTimeout time.Duration) {
+	timer := time.NewTimer(p.nextRefreshHint.Load().Sub(time.Now()))
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			klog.InfoS("Key cache shutting down")
+			return
+		case <-timer.C:
+		}
+
+		timedCtx, cancel := context.WithTimeout(ctx, keySyncTimeout)
+		if err := p.syncKeys(timedCtx); err != nil {
+			klog.Errorf("when syncing supported public keys(Stale set of keys will be supported): %v", err)
+			timer.Reset(fallbackRefreshDuration)
+		} else {
+			timer.Reset(p.nextRefreshHint.Load().Sub(time.Now()))
+			p.broadcastUpdate()
+		}
+		cancel()
+	}
+}
+
+func (p *keyCache) AddListener(listener externaljwt.Listener) {
+	p.listenersLock.Lock()
+	defer p.listenersLock.Unlock()
+
+	p.listeners = append(p.listeners, listener)
+}
+
+func (p *keyCache) GetCacheAgeMaxSeconds() int {
+	val := int(p.nextRefreshHint.Load().Sub(time.Now()).Seconds())
+	if val < 0 {
+		return 0
+	}
+	return val
+}
+
+// GetPublicKeys returns the public key corresponding to requested keyID.
+// Getter is expected to return All keys for keyID ""
+func (p *keyCache) GetPublicKeys(ctx context.Context, keyID string) []externaljwt.PublicKey {
+	pubKeys, ok := p.findKeyForKeyID(keyID)
+	if ok {
+		return pubKeys
+	}
+
+	// If we didn't find it, trigger a sync.
+	if err := p.syncKeys(ctx); err != nil {
+		klog.ErrorS(err, "Error while syncing keys")
+		return []externaljwt.PublicKey{}
+	}
+
+	pubKeys, ok = p.findKeyForKeyID(keyID)
+	if ok {
+		return pubKeys
+	}
+
+	// If we still didn't find it, then it's an unknown keyID.
+	klog.Errorf("Key id %q not found after refresh", keyID)
+	return []externaljwt.PublicKey{}
+}
+
+func (p *keyCache) findKeyForKeyID(keyID string) ([]externaljwt.PublicKey, bool) {
+	if len(p.publicKeys.Load().Keys) == 0 {
+		klog.Error("No keys currently in cache. Initial fill has not completed")
+		return nil, false
+	}
+
+	if keyID == "" {
+		return p.publicKeys.Load().Keys, true
+	}
+
+	for _, key := range p.publicKeys.Load().Keys {
+		if key.KeyID == keyID {
+			return []externaljwt.PublicKey{key}, true
+		}
```

Comment:
Uh I thought we could not enforce key ID uniqueness?

Reply:
Fixed to return all keys with KeyID